# Attempting to Unnest the Beast
It's really not that nested but I'm STRUGGLING so let's try this in Python. Here's my thought process:

* Turn json file into a dictionary
* Turn dictionary into pandas data frame
* Turn (hopefully neat) data frame into a csv file to be imported in R!

In [84]:
import pandas as pd
import json

In [85]:
f = open('polyvore_data/train_no_dup.json')
dict = json.load(f)
f.close()
df = pd.json_normalize(dict)

In [86]:
df.head()
df['items'].loc[1,]

[{'index': 1,
  'name': 'nirvana distressed t-shirt',
  'price': 10.0,
  'likes': 1290,
  'image': 'http://img2.polyvoreimg.com/cgi/img-thing?.out=jpg&size=m&tid=108042640',
  'categoryid': 21},
 {'index': 2,
  'name': 'rag bone rock w/ black skinny jeans',
  'price': 235.0,
  'likes': 17406,
  'image': 'http://img2.polyvoreimg.com/cgi/img-thing?.out=jpg&size=m&tid=103135174',
  'categoryid': 237},
 {'index': 3,
  'name': 'vans authentic black mono trainers',
  'price': 43.0,
  'likes': 10734,
  'image': 'http://img2.polyvoreimg.com/cgi/img-thing?.out=jpg&size=m&tid=107307153',
  'categoryid': 49},
 {'index': 4,
  'name': 'time low rubber bracelet hot topic',
  'price': 5.6,
  'likes': 373,
  'image': 'http://img2.polyvoreimg.com/cgi/img-thing?.out=jpg&size=m&tid=107349261',
  'categoryid': 106},
 {'index': 5,
  'name': 'veil logo rubber bracelet',
  'price': 7.0,
  'likes': 469,
  'image': 'http://img1.polyvoreimg.com/cgi/img-thing?.out=jpg&size=m&tid=106473878',
  'categoryid': 106},

In [87]:
name_list = []
category_list = []
for item in df['items'].loc[1,]:
    name_list.append(item['name'])
    category_list.append(item['categoryid'])

name_list
category_list # PERIODT!!!!


[21, 237, 49, 106, 106, 106, 106, 106]

In [88]:
name_list = []
category_list = []
for item in df['items'].loc[2,]:
    name_list.append(item['name'])
    category_list.append(item['categoryid'])

# max length of items is 8
if len(name_list) < 8:
    fill = 8 - len(name_list)
    name_list = name_list + (['empty'] * fill)
    category_list = category_list + ([0] * fill)
name_list
category_list

# this bit also works. now we need to find a way to add
# these as columns into the data frame

[104, 25, 237, 261, 37, 55, 0, 0]

We're going to want to create a list of lists so we can turn it into a dataframe. The lists should be n entries by 16 columns. We'll demonstrate this with a simple 2x16 list of lists (array?).

In [89]:
name_list1 = []
category_list1 = []
for item in df['items'].loc[1,]:
    name_list1.append(item['name'])
    category_list1.append(item['categoryid'])

# max length of items is 8
if len(name_list) < 8:
    fill = 8 - len(name_list1)
    name_list1 = name_list1 + (['empty'] * fill)
    category_list1 = category_list1 + ([0] * fill)
row1 = name_list1 + category_list1

name_list2 = []
category_list2 = []
for item in df['items'].loc[2,]:
    name_list2.append(item['name'])
    category_list2.append(item['categoryid'])

# max length of items is 8
if len(name_list2) < 8:
    fill = 8 - len(name_list2)
    name_list2 = name_list2 + (['empty'] * fill)
    category_list2 = category_list2 + ([0] * fill)
row2 = name_list2 + category_list2

data = [row1, row2]

df_test = pd.DataFrame(data, columns = ['iname1', 'iname2', 'iname3', 'iname4',
                                        'iname5', 'iname6', 'iname7', 'iname8',
                                        'iid1', 'iid2', 'iid3', 'iid4',
                                        'iid5', 'iid6', 'iid7', 'iid8'])

df_test



iname1                               iname2  \
0  nirvana distressed t-shirt  rag bone rock w/ black skinny jeans   
1               monki singlet                     joy denim jacket   

                                     iname3  \
0        vans authentic black mono trainers   
1  topshop moto joni high rise skinny jeans   

                               iname4                         iname5  \
0  time low rubber bracelet hot topic      veil logo rubber bracelet   
1         black pointed chelsea boots  pre-owned chanel shoulder bag   

                        iname6       iname7  \
0    rubber bracelet hot topic  romance i'm   
1  rag bone floppy brim fedora        empty   

                                              iname8  iid1  iid2  iid3  iid4  \
0  disney alice wonderland cat rubber bracelet ho...    21   237    49   106   
1                                              empty   104    25   237   261   

   iid5  iid6  iid7  iid8  
0   106   106   106   106  
1    37    55     0     0

**WAHOO!**
Now we need to make it a function. I'm sure I'll run into some error when I try to use all of the data lol.

In [90]:
cols_df = df['items'].loc[1:2,]

def itemscol_to_dataframe(column):
    data = []

    for index, sublist in column.items():
        name_list = []
        category_list = []

        for item in sublist:
            name_list.append(item['name'])
            category_list.append(item['categoryid'])

        if len(name_list) < 8:
            fill = 8 - len(name_list)
            name_list = name_list + (['empty'] * fill)
            category_list = category_list + ([0] * fill)
        row = name_list + category_list
        data.append(row)
        
    itemsdf = pd.DataFrame(data, columns = ['iname1', 'iname2', 'iname3', 'iname4',
                                        'iname5', 'iname6', 'iname7', 'iname8',
                                        'iid1', 'iid2', 'iid3', 'iid4',
                                        'iid5', 'iid6', 'iid7', 'iid8'])
    return itemsdf

itemscol_to_dataframe(cols_df)

iname1                               iname2  \
0  nirvana distressed t-shirt  rag bone rock w/ black skinny jeans   
1               monki singlet                     joy denim jacket   

                                     iname3  \
0        vans authentic black mono trainers   
1  topshop moto joni high rise skinny jeans   

                               iname4                         iname5  \
0  time low rubber bracelet hot topic      veil logo rubber bracelet   
1         black pointed chelsea boots  pre-owned chanel shoulder bag   

                        iname6       iname7  \
0    rubber bracelet hot topic  romance i'm   
1  rag bone floppy brim fedora        empty   

                                              iname8  iid1  iid2  iid3  iid4  \
0  disney alice wonderland cat rubber bracelet ho...    21   237    49   106   
1                                              empty   104    25   237   261   

   iid5  iid6  iid7  iid8  
0   106   106   106   106  
1    37    55     0     0

*IT WORKS ON OUR TEST SERIES!!!*

Now for the real test...

In [91]:
itemscol = df['items']
itemscoldf = itemscol_to_dataframe(itemscol)
itemscoldf

iname1  \
0              mock neck embroidery suede sweatshirt   
1                         nirvana distressed t-shirt   
2                                      monki singlet   
3                                               tops   
4      yoins leather sexy v-neck sleeveless crop top   
...                                              ...   
17311                 raey deep v-neck silk cami top   
17312           michael kors printed bow shirt world   
17313       color block fit flare maxi chiffon dress   
17314                      patchwork fit flare dress   
17315   burberry kensington leather trim trench coat   

                                                  iname2  \
0                          luxe double zip hooded jacket   
1                    rag bone rock w/ black skinny jeans   
2                                       joy denim jacket   
3      saint laurent zip cutout stretch nappa leather...   
4               solid color long sleeve irregular blazer   
...                                                  ...   
17311                     bridge patchwork leather skirt   
17312                        tweed classic blazer chanel   
17313          plaid pattern lapel neck long sleeve coat   
17314             belted high collar coat leather sleeve   
17315                            lace top chunky booties   

                                                  iname3  \
0            citizens humanity high rise rocket hem jean   
1                     vans authentic black mono trainers   
2               topshop moto joni high rise skinny jeans   
3      corset super store women's black steampunk corset   
4               alice+olivia floral pattern a-line skirt   
...                                                  ...   
17311                       mango pointed toe flat shoes   
17312                        pink tartan high rise pants   
17313                        isabel marant suede sandals   
17314                          gucci snake leather pumps   
17315      christian louboutin kate patent-leather pumps   

                                                  iname4  \
0                                  suede tie short boots   
1                     time low rubber bracelet hot topic   
2                            black pointed chelsea boots   
3      allurez square diamond halo engagement ring we...   
4                              zipped top chunky booties   
...                                                  ...   
17311       givenchy antigona pouch medium leather pouch   
17312                   balenciaga v leather ankle boots   
17313                                       shoulder bag   
17314          metal rhinestones pu leather shoulder bag   
17315             jimmy choo women glitter fabric clutch   

                                            iname5  \
0                     cloth travel school backpack   
1                        veil logo rubber bracelet   
2                    pre-owned chanel shoulder bag   
3           lip buckled matte womens corset lip uk   
4         gold boho turquoise leaf tassel earrings   
...                                            ...   
17311   women's charcoal round metal rx eyeglasses   
17312                     charles leather tote bag   
17313                                 pieces rings   
17314  bottega veneta eyewear square frame glasses   
17315                                 mac lipstick   

                                  iname6  \
0                                          
1              rubber bracelet hot topic   
2            rag bone floppy brim fedora   
3                     nude pink lipstick   
4             etro heart locket necklace   
...                                  ...   
17311   sigma travel kit make classy kit   
17312                              empty   
17313               double slit ring set   
17314  parker black gold trim medium pen   
17315                      ysl black edp   

                                

I feel like crying. I'm so glad that works lol. SHOCKED, really. I'm gonna pass out. Some of the entries are blank so I just want to see what kind of object they're stored as to see how I should handle this, or if I can convert it to 'empty'. Then I will convert it to 'empty'.

In [92]:
itemscoldf['iname7'].iloc[17315,] # it's just an empty string
itemscoldf.loc[itemscoldf['iname7'] == '', 'iname7'] = 'empty'
itemscoldf.iloc[17315,] # works for that one empty string I saw, so it will probably work for the others

iname1     burberry kensington leather trim trench coat
iname2                          lace top chunky booties
iname3    christian louboutin kate patent-leather pumps
iname4           jimmy choo women glitter fabric clutch
iname5                                     mac lipstick
iname6                                    ysl black edp
iname7                                            empty
iname8                                            empty
iid1                                                 24
iid2                                                261
iid3                                                 43
iid4                                                 38
iid5                                                200
iid6                                                 95
iid7                                                 11
iid8                                                  0
Name: 17315, dtype: object

In [119]:
itemscoldf.loc[itemscoldf.isin(['']).any(axis=1)].shape[0]

1038

There are 1038 rows that have an empty string in one of their item names' categories. 

In [120]:
itemscoldf.loc[itemscoldf['iname1'] == '', 'iname1'] = 'empty'
itemscoldf.loc[itemscoldf['iname2'] == '', 'iname2'] = 'empty'
itemscoldf.loc[itemscoldf['iname3'] == '', 'iname3'] = 'empty'
itemscoldf.loc[itemscoldf['iname4'] == '', 'iname4'] = 'empty'
itemscoldf.loc[itemscoldf['iname5'] == '', 'iname5'] = 'empty'
itemscoldf.loc[itemscoldf['iname6'] == '', 'iname6'] = 'empty'
itemscoldf.loc[itemscoldf['iname8'] == '', 'iname8'] = 'empty'
itemscoldf.loc[itemscoldf.isin(['']).any(axis=1)].shape[0]

0

Now there are zero.